### PR TITLE
[Patch] Support INDEX.LIST in jar file to speed up resource finding

### PR DESCRIPTION
**Pull request by kelthuzadx**
`jdk.internal.util.jar.JarIndex` was introduced by JDK1.3, it's an old but useful technique. If third-party JAR files contain INDEX.LIST file(generated via `jar -i` command), these JAR files can be identified as loading into JVM. Class loader such as BuiltinClassLoader,URLClassLoader and subclasses of URLClassLoader can speed up their resource searching with the help of INDEX.LIST since it maps resource names to corresponding JAR names. 

The latest version of Tomcat does not aware of `INDEX.LIST`, when there are so many JAR files in `WEB-INF/lib`,  Tom cat have to search in all JAR files sequentially, which is obviously slow. This patch helps Tomcat to be aware of `INDEX.LIST`, speeds up the search of resources, especially when there are many JAR files in `WEB-INF/lib`

### CODE REVIEW

File: java/org/apache/tomcat/util/scan/JarIndex.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package org.apache.tomcat.util.scan;
```
Comment (markt-asf) [inline java/org/apache/tomcat/util/scan/JarIndex.java:1]: No ALv2 Header
Why is this class in this package when it is only used in `org.apache.catalina.webresources`?

File: java/org/apache/catalina/WebResourceSet.java [hunk 2]
```diff
@@ -36,6 +37,10 @@
      */
     WebResource getResource(String path);
 
+    default WebResource getResource(String path, List<List<WebResourceSet>> allResources){
```
Comment (markt-asf) [inline java/org/apache/catalina/WebResourceSet.java:40]: This looks wrong. A single `WebResourceSet` shouldn't need to be passed a reference to `allResources`.

File: java/org/apache/catalina/webresources/JarResourceSet.java [hunk 3]
```diff
@@ -59,6 +63,36 @@ public JarResourceSet(WebResourceRoot root, String webAppMount, String base,
...
+        }
+        String webAppMount = getWebAppMount();
+        if (path.startsWith(getWebAppMount())) {
+            String pathInJar = getInternalPath() + path.substring(
+                    webAppMount.length());
+            if (pathInJar.length() > 0 && pathInJar.charAt(0) == '/') {
+                pathInJar = pathInJar.substring(1);
+            }
+            List<String> jarFiles = jarIndex.get(pathInJar);
+            if(jarFiles!=null){
```
Comment (markt-asf) [inline java/org/apache/catalina/webresources/JarResourceSet.java:80]: Why is this here rather than in `StandardRoot` ?

**Comment by kelthuzadx** [comment]
I will fix these problems and give a fresh new patch, anyway, thanks to your review.
